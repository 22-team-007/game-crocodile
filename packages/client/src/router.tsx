import { redirect, RouterProvider, createBrowserRouter } from 'react-router-dom'
import type { RouteObject, LoaderFunctionArgs } from 'react-router-dom'
// components
import App from './layouts/app/App'

import {
  StartPage,
  ErrorPage,
  Profile,
  PrivateProfile,
  Login,
  LeaderBoard,
  leaderBoardLoader,
  Game,
  GameList,
  Forum,
  Registration,
  Page,
  ForumTheme,
} from './pages'

import api from './api'
import { logoutUser } from './store/actions/user'
import { useAppDispatch } from './hooks/useAppSelector'
import { UserLogoutAction, UserDataAction } from './store/actions/types'
import { userTypes } from './store/actions/user'

export enum Routes {
  Index = '/',
  Login = 'signin',
  Register = 'signup',
  Logout = 'signout',
  GameList = 'game',
  Game = 'game/:chatId',
  Profile = 'profile',
  ProfileId = 'profile/:profileId',
  Leaderboard = 'leaders',
  Forum = 'forum',
  ForumTheme = 'forum/:id',
  E404 = '404',
  E500 = '500',
  OAuth = 'oauth',
}

export function getRouterConf(forTest = '') {
  let dispatch: (arg0: UserLogoutAction | UserDataAction) => any
  let OAuthLoader: ((arg: LoaderFunctionArgs) => any) | undefined

  if (!forTest) {
    dispatch = useAppDispatch()

    OAuthLoader = async ({ request }) => {
      const code = new URL(request.url).searchParams.get('code')

      if (code) {
        const redirectURI = '/oauth'

        let resp = await api.oauth.signIn(code, redirectURI)

        // try log out and enter again
        if (resp.reason === 'User already in system') {
          await api.auth.logOut()
          resp = await api.oauth.signIn(code, redirectURI)
        }

        if (resp.reason === 'ok') {
          const user = await api.auth.user()

          if (user?.id) {
            dispatch({ type: userTypes.SET_USER_DATA, payload: user })
          }
        }
      }
      return redirect('/')
    }
  } else {
    OAuthLoader = undefined
  }

  const routerConf: RouteObject[] = [
    {
      path: Routes.Index,
      element: <App />,
      errorElement: <ErrorPage />,
      children: [
        {
          index: true,
          element: (
            <Page title="Крокодил - Главная страница">
              <StartPage />
            </Page>
          ),
        },
        {
          path: Routes.Login,
          element: (
            <Page title="Крокодил - Вход">
              <Login />
            </Page>
          ),
        },
        {
          path: Routes.GameList,
          element: (
            <Page title="Крокодил - Игра">
              <GameList />
            </Page>
          ),
        },
        {
          path: Routes.Game,
          element: (
            <Page title="Крокодил - Игра">
              <Game />
            </Page>
          ),
        },
        {
          path: Routes.Register,
          element: (
            <Page title="Крокодил - Регистрация">
              <Registration />
            </Page>
          ),
        },
        {
          path: Routes.Profile,
          element: (
            <Page title="Крокодил - Профиль">
              <PrivateProfile />
            </Page>
          ),
        },
        {
          path: Routes.ProfileId,
          loader: undefined,
          element: (
            <Page title="Крокодил - Профиль">
              <Profile />
            </Page>
          ),
        },
        {
          path: Routes.Leaderboard,
          loader: leaderBoardLoader,
          element: (
            <Page title="Крокодил - Лидеры">
              <LeaderBoard />
            </Page>
          ),
        },
        {
          path: Routes.Forum,
          element: (
            <Page title="Крокодил - Форум">
              <Forum />
            </Page>
          ),
        },
        {
          path: Routes.ForumTheme,
          element: (
            <Page title="Крокодил - Форум">
              <ForumTheme />
            </Page>
          ),
        },
        {
          path: Routes.Logout,
          loader: async () => {
            await api.auth.logOut()
            forTest ? false : dispatch(logoutUser())
            return redirect('/')
          },
        },
      ],
    },
    {
      path: Routes.E404,
      element: (
        <Page title="Ошибка - 404">
          <ErrorPage />
        </Page>
      ),
    },
    {
      path: Routes.E500,
      element: (
        <Page title="Ошибка - 500">
          <ErrorPage />
        </Page>
      ),
    },
    {
      path: Routes.OAuth,
      loader: OAuthLoader,
    },
  ]

  return routerConf
}

export function Index() {
  const router = getRouterConf()

  return <RouterProvider router={createBrowserRouter(router)} />
}
