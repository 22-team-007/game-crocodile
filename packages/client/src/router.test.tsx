import { render, screen } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import { createMemoryRouter, RouterProvider } from 'react-router-dom'
import { getRouterConf } from './router'

// @ts-ignore
document.fullscreenElement = jest.fn(() => Promise.resolve())

const MockedNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  Navigate: (props: any) => {
    MockedNavigate(props.to)
    return
  },
}))

let userId = 1

jest.mock('./hooks/useAppSelector', () => ({
  useAppSelector: () => {
    return userId
  },
}))

jest.mock('./utils/sound', () => ({
  fullScreenIn: () => {
    return
  },
  fullScreenOut: () => {
    return
  },
}))

const Testconf = [...getRouterConf('forTest')]

describe('Testing router ', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  test('Get index page', async () => {
    const router = createMemoryRouter(Testconf, {
      initialEntries: ['/'],
    })

    await act(() => {
      render(<RouterProvider router={router} />)
    })

    const element = screen.getByText('Правила игры')
    expect(element).toMatchSnapshot()
  })

  test('Private route with registered user', async () => {
    //user is registerd id - 1
    userId = 1

    // @ts-ignore
    global.fetch = jest.fn(() =>
      // fetch empty list of game
      Promise.resolve({ json: () => Promise.resolve([]) })
    )

    const router = createMemoryRouter(Testconf, {
      initialEntries: ['/game'],
    })

    await act(() => {
      render(<RouterProvider router={router} />)
    })

    const element = screen.getByText('Создать игру')
    expect(element).toMatchSnapshot()
  })

  test('Private route with unregistered user', async () => {
    //user is unregistered id - 0
    userId = 0

    const router = createMemoryRouter(Testconf, {
      initialEntries: ['/game'],
    })

    await act(() => {
      render(<RouterProvider router={router} />)
    })

    expect(MockedNavigate).toHaveBeenCalledWith('/signin')
  })
})
