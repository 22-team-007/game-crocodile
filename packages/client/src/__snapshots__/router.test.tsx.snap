// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing router  Get index page 1`] = `
<h2
  id="rules"
>
  Правила игры
</h2>
`;

exports[`Testing router  Private route with registered user 1`] = `
<button
  class="btn btn-success"
  type="button"
>
  Создать игру
</button>
`;
